File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/bloomfilter/BloomFilter.scala
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

package org.opensearch.flint.spark.skipping.bloomfilter

import java.io.{InputStream, OutputStream}
import java.util.Locale

import org.opensearch.flint.spark.skipping.bloomfilter.BloomFilter.Algorithm.{Algorithm, CLASSIC}

trait BloomFilter {

  def algorithm: Algorithm

  def bitSize(): Long

  def put(item: Long): Boolean

  def merge(bloomFilter: BloomFilter): BloomFilter

  def mightContain(item: Long): Boolean

  def writeTo(out: OutputStream): Unit
}

object BloomFilter {

  object Algorithm extends Enumeration {
    type Algorithm = Value
    val CLASSIC = Value
  }

  val BLOOM_FILTER_ALGORITHM_KEY = "algorithm"
  val DEFAULT_BLOOM_FILTER_ALGORITHM = CLASSIC.toString

  class BloomFilterFactory(params: Map[String, String]) extends Serializable {

    private val algorithm: Algorithm = {
      val param = params.getOrElse(BLOOM_FILTER_ALGORITHM_KEY, DEFAULT_BLOOM_FILTER_ALGORITHM)
      Algorithm.withName(param.toUpperCase(Locale.ROOT))
    }

    def parameters: Map[String, String] = {
      algorithm match {
        case CLASSIC => ClassicBloomFilter.getParameters(params) // TODO: add algorithm param
      }
    }

    def create(): BloomFilter = {
      algorithm match {
        case CLASSIC => new ClassicBloomFilter(parameters)
      }
    }

    def deserialize(in: InputStream): BloomFilter = {
      algorithm match {
        case CLASSIC => ClassicBloomFilter.deserialize(in)
      }
    }
  }
}

File: flint-core/src/main/java/org/opensearch/flint/core/field/bloomfilter/BloomFilter.java
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

package org.opensearch.flint.core.field.bloomfilter;

import java.io.OutputStream;

/**
 * Bloom filter interface inspired by [[org.apache.spark.util.sketch.BloomFilter]] but adapts to
 * Flint skipping index use and remove unnecessary API for now.
 */
public interface BloomFilter {

  /**
   * @return the number of bits in the underlying bit array.
   */
  long bitSize();

  /**
   * Put an item into this bloom filter.
   *
   * @param item Long value item to insert
   * @return true if bits changed which means the item must be first time added to the bloom filter.
   * Otherwise, it maybe the first time or not.
   */
  boolean put(long item);

  /**
   * Merge this bloom filter with another bloom filter.
   *
   * @param bloomFilter bloom filter to merge
   * @return bloom filter after merged
   */
  BloomFilter merge(BloomFilter bloomFilter);

  /**
   * @param item Long value item to check
   * @return true if the item may exist in this bloom filter. Otherwise, it is definitely not exist.
   */
  boolean mightContain(long item);

  /**
   * Serialize this bloom filter and write it to an output stream.
   *
   * @param out output stream to write
   */
  void writeTo(OutputStream out);
}
